includes/profile.inc

```diff
@@ -449,3 +459,368 @@ function cm_tools_update_missing_default_field_values_insert_with_languages(arra
+ *   - $table_type - One of CM_TOOLS_FIELD_TABLE_DATA, or
+ *                   CM_TOOLS_FIELD_TABLE_REVISION.
+ *   The $work_callback can return one of these values:
+ *   - NULL (default): Updates the row if a column was changed. Changes to
+ *                     columns used in the primary key are ignored.
+ *   - TRUE: Forces the row to be updated, even if nothing changed,
+ *   - FALSE: Skips row update, even if something changed.
+ * @param int $rows_per_pass
+ *   The number of table rows to load and process in each batch iteration.
+ * @param int $tables
+ *   Integer flags indicating which field tables to include. Defaults to both
+ *   'data' and 'revision' tables. Possible values are:
+ *   - CM_TOOLS_FIELD_TABLE_DATA -- include field data tables only,
+ *   - CM_TOOLS_FIELD_TABLE_REVISION -- include field revision tables only.
+ *
+ * @return string
+ *   A status message suitable for the return value of your update hook.
+ */
+function cm_tools_field_tables_walk(array &$sandbox, callable $field_list_filter_callback, callable $query_alter_callback, callable $work_callback, $rows_per_pass = 50, $tables = CM_TOOLS_FIELD_TABLE_DATA | CM_TOOLS_FIELD_TABLE_REVISION) {
+  // Convenience reference.
+  if (!isset($sandbox['_cm_tools_field_tables_walk'])) {
+    $sandbox['_cm_tools_field_tables_walk'] = array();
+  }
+  $s = &$sandbox['_cm_tools_field_tables_walk'];
+
+  $status = array();
+
+  // Initialize state, shared variables and gather fields.
+  if (empty($s)) {
+    $s['fields'] = call_user_func(
+      $field_list_filter_callback,
+      // We only support fields using the default SQL storage.
+      array_filter(field_info_fields(), function ($field) {
+        return $field['storage']['module'] == 'field_sql_storage' &&
+          $field['storage']['type'] == 'field_sql_storage';
+      }),
+      field_info_instances()
+    );
+    $s['fields_count'] = count($s['fields']);
+    $s['field'] = NULL;
+    $s['table_type'] = 0;
+    $tables_str = array();
+    if ($tables & CM_TOOLS_FIELD_TABLE_DATA) {
+      $tables_str[] = 'data';
+    }
+    if ($tables & CM_TOOLS_FIELD_TABLE_REVISION) {
+      $tables_str[] = 'revision';
+    }
+    $tables_str = implode($tables_str, ' and ') . (count($tables_str) > 1 ? ' tables' : ' table');
+    $status[] = strtr('Processing the !tables for !field_count fields.', array(
+      '!tables' => $tables_str,
+      '!field_count' => $s['fields_count'],
+    ));
+  }
+
+  $status[] = cm_tools_tables_walk($sandbox,
+    // Move to next table callback.
+    function () use (&$s, $tables) {
+      if ($tables === CM_TOOLS_FIELD_TABLE_DATA | CM_TOOLS_FIELD_TABLE_REVISION) {
+        if ($s['table_type'] === CM_TOOLS_FIELD_TABLE_DATA) {
+          $s['table_type'] = CM_TOOLS_FIELD_TABLE_REVISION;
+        }
+        else {
+          $s['field'] = array_shift($s['fields']);
+          $s['table_type'] = CM_TOOLS_FIELD_TABLE_DATA;
+        }
+      }
+      else {
```

Comment:
What is this else case intended to handle? Surely the `$tables` variable should only ever be `CM_TOOLS_FIELD_TABLE_DATA` or `CM_TOOLS_FIELD_TABLE_REVISION`, in which case `($tables === CM_TOOLS_FIELD_TABLE_DATA | CM_TOOLS_FIELD_TABLE_REVISION)` would always be true?

Reply:
So (warning) that's a single pipe-character, so it's doing a bitwise OR. This is all FLAG logic. So `$tables === CM_TOOLS_FIELD_TABLE_DATA | CM_TOOLS_FIELD_TABLE_REVISION)` is only true when the value is `00000011` and so BOTH the data and revision tables have been specified.

The else case here is intended to handle the situation when the `$tables` is either `00000001` or `00000010` (so either one of the data and revision tables (but not both) have been specified).

Reply:
OK, I think I'd got a bit confused by the bitwise operation. I had spotted it, but misunderstood it.